core/include/detray/core/detail/context_aware_store.hpp

```diff
@@ -0,0 +1,360 @@
+          typename context_t>
+class context_aware_store {
+    public:
+    using base_type = container_t<T>;
+    using vector_type = typename base_type::value_type;
+    using context_type = context_t;
+
+    // TODO: do we need this?
+    /*
+    /// How to find data in the store
+    /// @{
+    using link_type = dindex;
+    using single_link = dindex;
+    using range_link = dindex_range;
+    /// @}
+    */
+
+    /// Vecmem view types
+    using view_type = detail::get_view_t<container_t<T>>;
+    /*
+     TODO: uncommenting the next line results in compilation problems:
+     Originated from container_views.hpp
+     error: more than one partial specialization matches the template argument
+     list of class "detray::detail::has_view<.......>"
+     "detray::detail::has_view<const vecmem::vector<T>, void>"
+     "detray::detail::has_view<const vecmem::jagged_vector<T>, void>"
+    */
+    // using const_view_type = detail::get_view_t<const container_t<T>>;
+
+    /// Empty container
+    constexpr context_aware_store() = default;
+
+    // Delegate constructors to container, which handles the memory
+
+    /// Copy construct from element types
+    // TODO: do we need this?
+    /* constexpr explicit single_store(const T &arg) : m_container(arg) {} */
+
+    /// Construct with a specific memory resource @param resource
+    /// (host-side only)
+    template <typename allocator_t = vecmem::memory_resource,
+              std::enable_if_t<not detail::is_device_view_v<allocator_t>,
+                               bool> = true>
+    DETRAY_HOST explicit context_aware_store(allocator_t& resource)
+        : m_container(&resource) {}
+
+    /// Copy Construct with a specific memory resource @param resource
+    /// (host-side only)
+    // TODO: do we need this?
+    /*
+    template <typename allocator_t = vecmem::memory_resource,
+              typename C = container_t<T>,
+              std::enable_if_t<std::is_same_v<C, std::vector<T>>, bool> = true>
+    DETRAY_HOST explicit context_aware_store(allocator_t &resource, const T
+    &arg) : m_container(&resource, arg) {}
+    */
+
+    /// Construct from the container @param view . Mainly used device-side.
+    template <typename container_view_t,
+              std::enable_if_t<detail::is_device_view_v<container_view_t>,
+                               bool> = true>
+    DETRAY_HOST_DEVICE context_aware_store(container_view_t& view)
+        : m_container(view) {}
+
+    /// @returns a pointer to the underlying container - const
+    DETRAY_HOST_DEVICE
+    constexpr auto data() const noexcept -> const base_type* {
+        return &m_container;
+    }
+
+    /// @returns a pointer to the underlying container - non-const
+    DETRAY_HOST_DEVICE
+    constexpr auto data() noexcept -> base_type* { return &m_container; }
+
+    /// @returns the size of the underlying container
+    DETRAY_HOST_DEVICE
+    constexpr auto size(const context_type& ctx) const noexcept -> dindex {
```

Comment:
This is also for my understanding: will all of the contices have the same number of elements @asalzburger ?